Multiapp finalize

permcody opened the pull request:
This PR contains a bug fix and an enhancement to the output of Multiapps. First it fixes the Perf_graph output for multiapps, which wasn't really working since we weren't executing EXEC_FINAL objects at all in Multiapps. Secondly, it adds MAGENTA coloring to all Picard iteration outputs. 

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

permcody committed:
Properly print perf_graphs with multiapps

Perf Graphs with multiapps didn't quite work right. They would come out
accidentally when we got lucky and hit an outputStep(EXEC_FORCED) call
instead of actually honoring the usual EXEC_FINAL flag. This PR adds a
new finalize method to MultiApp that gets called after the multiapp
has finished executing. This method makes sure to execute all objects
with an EXEC_FINAL flag including outputs (such as PerfGraph).

refs #11551
closes #11453 #12186

permcody committed:
Testing for perf_graph outputs with MultiApps

refs #12186

permcody committed:
Add some color to make Picard iterations stand out

Also moving the outoutNorm function to public static so it can be used from other places.

permcody commented:
BTW - If people don't like the magenta, it can be ripped out. It's a separate commit in this PR.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/273375/) on bcd0e40 wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/12672/site)

This comment will be updated on new commits.

permcody commented:
@friedmud - You should probably review since this touches your code the most. I fixed up the Multiapp EXEC_FINAL stuff, changed the PerfGraph so that it prints a more informative "first" level (e.g. MooseTestApp (main), BisonTestApp(sub_app0), etc. In doing so, I made it so you could just ask for "Root" to get the first level for consistency so you don't have to know the exact format of that first entry. Finally, I added coloring to Multiapp Picard iteration simulations. Everything having to do with the Picard iteration comes out Magenta making it much easier to pick out during the run. This commit is separate if you don't like it.

friedmud commented on framework/src/utils/PerfGraph.C:
```diff
@@ -21,10 +21,13 @@
 {
   // Not done in the initialization list on purpose because this object needs to be complete first
-  _root_node = libmesh_make_unique<PerfNode>(registerSection("App", 0));
+  _root_node = libmesh_make_unique<PerfNode>(registerSection(ROOT_NAME, 0));
```
I really don't see a reason to change this.  Why is "Root" better than "App"?  "App" is what is printed in the graph... (and is already in use in some of my personal scripts - I'm sure others are using it too).

friedmud commented on framework/src/utils/PerfGraph.C:
```diff
@@ -21,10 +21,13 @@
 #include <chrono>
 
-PerfGraph::PerfGraph() : _current_position(0), _active(true)
+const std::string PerfGraph::ROOT_NAME = "Root";
```
I don't see a purpose in defining this - you didn't even use it in the error messages below (although I don't think it's necessary to use it in the error messages).

friedmud commented on framework/src/utils/PerfGraph.C:
```diff
@@ -74,7 +75,10 @@ PerfGraph::getNumCalls(const std::string & section_name)
+    mooseError(
+        "Unknown section_name: ",
+        section_name,
+        " in PerfGraph::getNumCalls()\nIf you are attempting to retrieve the root use \"Root\".");
```
This seems unnecessary.  The name you use is always what shows up in the output.  There are a million options but all you need to do is put exactly what's in the output an you're good.

friedmud commented on framework/src/utils/PerfGraph.C:
```diff
@@ -217,7 +224,8 @@ PerfGraph::recursivelyPrintGraph(PerfNode * current_node,
               "Unable to find section name!");
-  auto & name = _id_to_section_name[current_node->id()];
+
+  auto & name = current_node->id() == 0 ? _root_name : _id_to_section_name[current_node->id()];
```
Just why?

friedmud reviewed (changes requested):
This is all good to me - but I don't see a purpose in the "Root" change.  I like the "App".  It's the total "App Time".  But whatevs...

permcody commented:
All of your comments are related to the whole root change, so let me explain what changed. I was working on printing out the perf_graph and I got it to come out first but it wasn't coming out on the console. I was getting two logs (one for the main, one for the sub) but it was pretty tricky to figure out which was which. I eventually got it to come out with the console, but I hate relying on that to differentiate output all the time so I thought about customizing the output so you could see which app was which. Here's the current output with this PR:

```
sub_app0: 
sub_app0: Performance Graph:
sub_app0: ------------------------------------------------------------------------------------------------------------------------------------------------------------
sub_app0: |                 Section                | Calls |   Self(s)  |   Avg(s)   |    %   | Children(s) |   Avg(s)   |    %   |  Total(s)  |   Avg(s)   |    %   |
sub_app0: ------------------------------------------------------------------------------------------------------------------------------------------------------------
sub_app0: | MooseTestApp (sub_app0)                |     1 |      0.090 |      0.090 |   8.18 |       1.011 |      1.011 |  91.82 |      1.101 |      1.101 | 100.00 |
sub_app0: |   FEProblem::solve                     |     5 |      0.056 |      0.011 |   5.12 |       0.862 |      0.172 |  78.26 |      0.918 |      0.184 |  83.38 |
sub_app0: |     FEProblem::computeResidualInternal |    80 |      0.000 |      0.000 |   0.00 |       0.705 |      0.009 |  64.03 |      0.705 |      0.009 |  64.03 |
sub_app0: |     FEProblem::computeJacobianInternal |    10 |      0.000 |      0.000 |   0.00 |       0.156 |      0.016 |  14.17 |      0.156 |      0.016 |  14.17 |
sub_app0: |   FEProblem::outputStep                |    12 |      0.000 |      0.000 |   0.02 |       0.008 |      0.001 |   0.73 |      0.008 |      0.001 |   0.75 |
sub_app0: ------------------------------------------------------------------------------------------------------------------------------------------------------------

Performance Graph:
------------------------------------------------------------------------------------------------------------------------------------------------------------
|                 Section                | Calls |   Self(s)  |   Avg(s)   |    %   | Children(s) |   Avg(s)   |    %   |  Total(s)  |   Avg(s)   |    %   |
------------------------------------------------------------------------------------------------------------------------------------------------------------
| MooseTestApp (main)                    |     1 |      0.012 |      0.012 |   1.01 |       1.174 |      1.174 |  98.99 |      1.186 |      1.186 | 100.00 |
|   FEProblem::solve                     |     5 |      0.009 |      0.002 |   0.75 |       0.063 |      0.013 |   5.34 |      0.072 |      0.014 |   6.10 |
|     FEProblem::computeResidualInternal |    77 |      0.000 |      0.000 |   0.00 |       0.052 |      0.001 |   4.40 |      0.052 |      0.001 |   4.40 |
|     FEProblem::computeJacobianInternal |    10 |      0.000 |      0.000 |   0.00 |       0.011 |      0.001 |   0.89 |      0.011 |      0.001 |   0.90 |
|   FEProblem::outputStep                |    11 |      0.000 |      0.000 |   0.02 |       0.001 |      0.000 |   0.06 |      0.001 |      0.000 |   0.08 |
|   FEProblem::execMultiApps             |     5 |      0.927 |      0.185 |  78.23 |       0.000 |      0.000 |   0.00 |      0.927 |      0.185 |  78.23 |
|   Transient::final                     |     1 |      0.000 |      0.000 |   0.02 |       0.000 |      0.000 |   0.00 |      0.000 |      0.000 |   0.02 |
|     FEProblem::outputStep              |     1 |      0.000 |      0.000 |   0.00 |       0.000 |      0.000 |   0.00 |      0.000 |      0.000 |   0.00 |
------------------------------------------------------------------------------------------------------------------------------------------------------------
```

"App" is now gone and it's clear right in the log which app or which sub this log belongs to. This is why I changed it to "Root" in reference to the whole tree. However, I didn't want users to get frustrated in obtaining the total time as they change their app name around or move it from the master to one of the subs or whatever. You don't want to have to put "section_name = "MooseTestApp (main)" in your input file and have that become brittle as you change things around. 

So let me know what you'd like here. If you just want to go back to "App", that's fine, do you like the update to the way the log prints?

friedmud commented:
Ok - sounds good.
